src/main/scala/com/fulcrumgenomics/bam/IdentifyPrimers.scala

```diff
@@ -0,0 +1,1027 @@
+    }
+    // Summary metrics
+    Metric.write(PathUtil.pathTo(this.metrics + ".summary.txt"), IdentifyPrimersMetric(metricCounter))
+  }
+
+  /** Creates a new [[InteractiveAlignmentScorer]]. */
+  private def newAligner: InteractiveAlignmentScorer[Primer, SamRecordAlignable] = {
+    InteractiveAlignmentScorer(matchScore, mismatchScore, gapOpen, gapExtend, AlignmentMode.Glocal, this.kswExecutable)
+  }
+
+  // NB: batching alignment inputs to the aligner (i.e. ksw) is empirically faster than given them all or just one at a time.
+  private def completePrimerMatch(matchOption: Option[Either[PrimerMatch, Seq[AlignmentTask[Primer, SamRecordAlignable]]]],
+                                  aligner: InteractiveAlignmentScorer[Primer, SamRecordAlignable],
+                                  alignmentInputBatchSize: Int = 100): Option[PrimerMatch] = {
+    case class PrimerAndScore(primer: Primer, score: Int)
+    import IdentifyPrimers.MaxTwoBy
+    matchOption.flatMap {
+      case Left(primerMatch)      => Some(primerMatch)
+      case Right(alignmentInputs) =>
+        // Add the alignment inputs to the aligner
+        alignmentInputs.grouped(alignmentInputBatchSize).foreach { batch =>
+          aligner.append(batch: _*)
+        }
+        // Get the alignment results
+        val alignmentResults = alignmentInputs.toIterator.zip(aligner.iterator)
+          // Keep results that meet the minimum score and align with a minimum # of query bases
+          .filter { case (alignmentInput, alignmentResult) =>
+            val minQueryEnd = alignmentInput.queryLength - slop
+            alignmentResult.score >= minAlignmentScore && alignmentResult.queryEnd >= minQueryEnd
+          }
+          // now just keep the primer and score
+          .map { case (alignmentInput, alignmentResult) => PrimerAndScore(alignmentInput.query, alignmentResult.score) }
+          // get the best two alignment results by score
+          .maxTwoBy(_.score)
+
+        // Create a primer match if alignments were found (i.e. a Option[PrimerMatch])
+        alignmentResults match {
+          case (None, _)                    => None
+          case (Some(best), None)           => Some(FullAlignmentPrimerMatch(primer = best.primer, score = best.score, secondBestScore = minAlignmentScore))
+          case (Some(best), Some(nextBest)) => Some(FullAlignmentPrimerMatch(primer = best.primer, score = best.score, secondBestScore = nextBest.score))
+        }
+    }
+  }
+
+  /** Processes a single batch of templates. */
+  def processBatch(templates: Seq[Template],
+                   metricCounter: SimpleCounter[TemplateTypeMetric],
+                   progress: ProgressLogger): Seq[Template] = {
+
+    val counter = new SimpleCounter[TemplateTypeMetric]()
+    val aligner  = newAligner
+
+    templates.foreach { template =>
+      // match up to but not including full alignment
+      val r1MatchOption = template.r1.map { r1 => fwdMatcher.matchUntilAlignment(r1) }
```

Comment:
Maybe just rename the variable in the closure to `r`.  It's confusing on the next line when you map `r2` and call it `r1`.